app/src/main/java/com/github/sdpteam15/polyevents/user/User.kt

```diff
@@ -0,0 +1,135 @@
+            if (firebaseUser == null)
+                throw IllegalArgumentException("FirebaseUser must be not null")
+            if (!instances.containsKey(firebaseUser.uid))
+                instances[firebaseUser.uid] = User(firebaseUser)
+            return instances[firebaseUser.uid] as User
+        }
+
+        /**
+         * Application user
+         * @param uid associates FirebaseUser.uid
+         * @return the application user associates to the uid
+         */
+        fun invoke(uid: String?): User? {
+            if (uid == null)
+                return null
+            return instances[uid]
+        }
+
+        private var lastCurrentUserUid: String? = null
+
+        /**
+         * Current user of the application
+         */
+        val CurrentUser: UserInterface?
+            get() {
+                try {
+                    val currentUser: FirebaseUser =
+                        FirebaseAuth.getInstance().currentUser ?: return null
+                    if (currentUser.uid != lastCurrentUserUid)
+                        invoke(lastCurrentUserUid)?.removeCache()
+                    return invoke(FirebaseUserAdapter(currentUser))
+                } catch (e: ExceptionInInitializerError) {
+                    return null
+                } catch (e: NoClassDefFoundError) {
+                    return null
+                }
+            }
+    }
+
+    private var profileList: MutableList<ProfileInterface>? = null
+    private var currentProfileId: Int = 0
+    var database: DatabaseInterface = DatabaseObject.Singleton
+
+    private fun FromDatabase(): MutableList<ProfileInterface> {
+        if (profileList == null)
+            profileList = database.getListProfile(UID, this).toMutableList()
+        if ((profileList as MutableList<ProfileInterface>).size == 0) {
+            val profile = Profile(Name)
+            (profileList as MutableList<ProfileInterface>).add(profile)
+            database.addProfile(profile, UID, this)
+        }
+        return profileList as MutableList<ProfileInterface>
+    }
+
+    override val ProfileList: List<ProfileInterface>
+        get() = FromDatabase() as List<ProfileInterface>
+
+    override val Name: String
+        get() = FirebaseUser.displayName
+    override val UID: String
+        get() = FirebaseUser.uid
+    override val Email: String
+        get() = FirebaseUser.email
+
+    /**
+     * Current application user profile id
+     */
+    var CurrentProfileId: Int
+        get() = currentProfileId
+        set(value) {
+            if (value < 0 || value >= ProfileList.size)
+                throw IndexOutOfBoundsException("value must be between 0 and ${ProfileList.size}")
+            else currentProfileId = value
+        }
+
+    override var CurrentProfile: ProfileInterface
+        get() = ProfileList[currentProfileId]
+        set(value) {
+            var i = 0
+            for (v in ProfileList) {
```

Comment:
(Detail : possible to use "ProfileList.withIndex()", which is equivalent to a enumerate in Python)